File: src/adapter/example/delegate/PrintBannerAdapter.java
package adapter.example.delegate;

import adapter.example.Banner;
import adapter.example.Printer;

public class PrintBannerAdapter implements Printer {
    private final Banner banner;

    public PrintBannerAdapter(Banner banner) {
        this.banner = banner;
    }

    @Override
    public void print() {
        System.out.println(banner.decorate());
    }
}

File: src/adapter/exercise2/delegate/PrintBannerAdapter.kt
package adapter.exercise2.delegate

import adapter.exercise2.Banner
import adapter.exercise2.Printer

class PrintBannerAdapter(val banner: Banner): Printer {
    override fun print() {
        println(banner.decorate())
    }
}

class PrintBannerAdapterBy(val banner: Banner): Printer, Banner by banner {
    override fun print() {
        println(decorate())
    }
}